### PR TITLE
[cudnn rnn fp32 math accumulation]

### PR DESCRIPTION
**Pull request by jjsjann123**
Summary: setting the correct math type for cudnn rnn, which is enforced starting from cudnn 7.5+

1. Updating persistent rnn check with input data type instead of rnn math type;
2. Updating rnn type promotion to set correct math type for accumulation;
3. Replace datatype check for filter descriptor from rnn.datatype to input.datatype;

### CODE REVIEW

File: aten/src/ATen/native/cudnn/RNN.cpp [hunk 1]
```diff
@@ -600,6 +600,15 @@ namespace {
 #endif
   }
 
+  cudnnDataType_t promote_rnn_math_type(cudnnDataType_t dtype) {
+#if CUDNN_VERSION != 7103
```
Comment (soumith) [inline aten/src/ATen/native/cudnn/RNN.cpp]: is it really `!=`? is it not something like `>=`? just checking

Comment (jjsjann123) [reply aten/src/ATen/native/cudnn/RNN.cpp]: yes this is supposed to be `!=`
In CUDNN 7.1.3 math type in rnn descriptor means something else and they have a check for it, so we are making an exception here.

**Comment by jjsjann123** [comment]
Ping @ngimel for visibility

File: aten/src/ATen/native/cudnn/RNN.cpp (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/native/cudnn/RNN.cpp]: This is one of those cases where having a comment explaining that yes, it's intentional, helps future code readers ;)

**Comment by ezyang** [comment]
I guess we're gonna need a cudnn 3.5 build soon...

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
@ngimel let me know if you want to review this before we put it in!

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**